pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,346 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/archive"
+	"github.com/vmware-tanzu/velero/pkg/client"
+	"github.com/vmware-tanzu/velero/pkg/features"
+)
+
+// meetsAPIGVRestoreReqs determines if the feature flag has been
+// enabled and the backup objects use Status.FormatVersion 1.1.0.
+func (ctx *restoreContext) meetsAPIGVRestoreReqs() bool {
+	return features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag) &&
+		ctx.backup.Status.FormatVersion >= "1.1.0"
+}
+
+// ChosenGroupVersion is the API Group version that was selected to restore
+// from potentially multiple backed up version enabled by the feature flag
+// APIGroupVersionsFeatureFlag
+type ChosenGroupVersion struct {
+	Group   string
+	Version string
+	Dir     string
+}
+
+// chooseAPIVersionsToRestore will choose a version to restore based on a user-
+// provided config map prioritization or our version prioritization.
+func (ctx *restoreContext) chooseAPIVersionsToRestore() error {
+	sgvs, tgvs, ugvs, err := ctx.gatherSTUVersions()
+	if err != nil {
+		return err
+	}
+
+OUTER:
+	for rg, sg := range sgvs {
+		// Default to the source preferred version if no other common version
+		// can be found.
+		cgv := ChosenGroupVersion{
+			Group:   sg.Name,
+			Version: sg.PreferredVersion.Version,
+			Dir:     sg.PreferredVersion.Version + "-preferredversion",
```

Comment:
We should use the constant PreferredVersionDir

Reply:
Good catch. I'll replace these with `velerov1api.PreferredVersionDir`.